### PR TITLE
15 ed exclusion flag for n imaging and n bloodwork

### LINKED ISSUE #15: Add flag to `n_imaging` and `n_bloodwork` to allow for exclusion of ED
## New Feature Request

Both functions currently include both ED and IP imaging/bloodwork tests. For certain projects (and MPR/OPR!), we only want to include IP tests. It would therefore be useful to add a flag `include_ed` that allows users to choose whether tests that were performed while patients were in ED should be included or not. The default of this flag should probably be set to `FALSE` although this will cause differences in results compared to previous versions of this function. 

For consistency, it might be nice to include a similar flag in the `n_rbc_transfusions` function, which currently only includes IP transfusions (i.e., transfusions before admitting date-time are excluded). Ideally, all 3 functions have the same flag following a similar logic.

To exclude tests performed during ED, we could merge with `ipadmdad` and exclude any tests that were performed prior to `admission_date_time`. This is the approach used in `n_rbc_transfusions`.

Note: The `n_imaging` function currently has a `dtvar` input argument. However, this argument is currently not used (but will be useful for exclusion of ED imaging).

### PR DESCRIPTION
**Pull request by guoyi-smh**
closes #15 

Refactored `n_imaging`, `n_routine_bloodwork` and `n_rbc_transfusion`. Now all functions take three arguments:
-  `dbcon`: database connection
-  `cohort`: patient list
-  `exclude_ed`:  a new logical argument to indicate whether tests in ED to be excluded.

Other changes include:
- All three functions will now print a note indicating that the function currently does not take data coverage into account and hence it is the user's responsibility to check data availability and interpret the 0s in the output table.
- When `exclude_ed` is set to `FALSE`, all tests are included, including tests without date time, tests before triage and after discharge
- `n_imaging` will no longer take a user specified date-time variable name, following implementation in MPR, `ordered_date_time` is used, when it is not available, `performed_date_time` is used instead to identify tests in ED. 
- `n_imaging` will not take a user-specified modality variable name, it will automatically use `modality_mapped` from GEMINI `radiology` table.
- For `n_rbc_transfusion`, following implementation in MPR, transfusions without a valid hemoglobin test in previous 48h are not counted.
- For `n_routine_bloodwork`, the filter to include only numeric results is now removed, to be consistent with definition in MPR and OPR. The rationale is to count all tests where blood was drawn from patients regardless of result values.
- `check_input` is now applied to all input arguments

### CODE REVIEW

**Review by loffleraSMH — COMMENTED** [review]
Thanks so much for implementing this Yishan! It’s nice to have this improved consistency across the 3 functions and also have the additional warning messages. 

I tested all functions to make sure they run as expected. For `n_rbc_transfusions`, the new output is a tiny bit different from before (when `exclude_ed = TRUE`) but I believe that’s simply due to the fact that the previous version excluded transfusions after discharge? For `n_routine_bloodwork` and `n_imaging`, the results are identical (if `exclude_ed = FALSE`) as expected.

I can do one more quick round of review once you’ve had a chance to address the comments above (no rush!).

**Comment by loffleraSMH** [comment]
@guoyi-smh one more general comment/question (for a future sprint): Some of the filtering by date-time variables happens within the SQL query now, which probably makes it more efficient, but with this approach it’s not obvious how many date-time variables are missing/invalid. I’ve been adding a new utility function that transforms/checks date-time variables (see issue #76), so I think eventually, I’d like to apply this within these 3 functions as well so users are informed about missing/invalid date-times? This might be particularly relevant for imaging, so we can inform users how missing order date-times are handled. That would mean we’d query all entries first and then perform the filtering by date-times after. Any thoughts on this approach? If you agree, I can open a new issue for that so we can tackle that in a future update.

File: R/n_imaging.r [hunk 1]
```diff
@@ -41,51 +39,83 @@
...
+    dbcon,
+    paste0(
+      ifelse(exclude_ed == TRUE,
+            # filter by admission date time and exclude tests before admission
+            paste0("with temp as (
+              select r.*,a.admission_date_time,
+              case when r.ordered_date_time is null or r.ordered_date_time = '' or
+              r.ordered_date_time = ' ' then r.performed_date_time >= a.admission_date_time
+              else r.ordered_date_time >= a.admission_date_time end as case_result
+              from radiology r
```
[the remark this answers is not in the document]
Comment (guoyi-smh) [reply R/n_imaging.r:102]: Thanks for catching this Anne and yes I hardcoded the table name here because `fin_db_tablename` does not allow `radiology` to be the input table name yet. You plan sounds great. Thanks!

File: R/utils.R [hunk 2]
```diff
@@ -621,11 +621,41 @@ check_input <- function(arginput, argtype,
...
+#' @param what (`character`)\cr
+#' Which clinical table is used.
+#'
+#' @param addtl (`character`)\cr
+#' An additional, specific message to append to the generic message.
+#'
+#' @return
+#' Prints a message to the console.
+#'
+availability_message <- function(what, addtl = NULL) {
```
[the remark this answers is not in the document]
Comment (guoyi-smh) [reply R/utils.R:649]: Agreed! Will update the function name and the msg.

File: R/n_rbc_transfusions.R [hunk 3]
```diff
@@ -69,105 +84,98 @@ n_rbc_transfusions <-function(dbcon,
   }
   cohort_subset <- cohort[hospital_num %ni% c(105, 106)]
```
[the remark this answers is not in the document]
Comment (guoyi-smh) [reply R/n_rbc_transfusions.R:85]: Thanks for catching this. Completely oversight on me. Will fix in next push.

File: R/n_imaging.r (hunk 1 above)
[the remark this answers is not in the document]
Comment (loffleraSMH) [reply R/n_imaging.r:102]: Perfect, I just pushed a commit that adds the radiology table to `find_db_tablename` so this should hopefully work now.

**Comment by guoyi-smh** [comment]
> Thanks so much for implementing this Yishan! It’s nice to have this improved consistency across the 3 functions and also have the additional warning messages.
> 
> I tested all functions to make sure they run as expected. For `n_rbc_transfusions`, the new output is a tiny bit different from before (when `exclude_ed = TRUE`) but I believe that’s simply due to the fact that the previous version excluded transfusions after discharge? For `n_routine_bloodwork` and `n_imaging`, the results are identical (if `exclude_ed = FALSE`) as expected.
> 
> I can do one more quick round of review once you’ve had a chance to address the comments above (no rush!).

I think you are right, the difference in `transfusion` should be caused by the filter on `discharge_date_time`. I dropped that for now to be consistent across 3 function. I had a conversation with Surain on this, and he suggested that in MPR we should have filtered by `discharge_date_time` for imaging too, which is currently not implemented. For the functions in `Rgemini`, by default, we would want to exclude records before `triage_date_time` and  after `discharge_date_time`. To add these filters I think it is better we address the coverage issue first, because we will have coverage issue for `er` table as well as the clinical table we are looking at.

**Comment by guoyi-smh** [comment]
> @guoyi-smh one more general comment/question (for a future sprint): Some of the filtering by date-time variables happens within the SQL query now, which probably makes it more efficient, but with this approach it’s not obvious how many date-time variables are missing/invalid. I’ve been adding a new utility function that transforms/checks date-time variables (see issue #76), so I think eventually, I’d like to apply this within these 3 functions as well so users are informed about missing/invalid date-times? This might be particularly relevant for imaging, so we can inform users how missing order date-times are handled. That would mean we’d query all entries first and then perform the filtering by date-times after. Any thoughts on this approach? If you agree, I can open a new issue for that so we can tackle that in a future update.

Agreed! I hesitated on this but chose the performance because I thought the numbers were not reported now anyways. As the user I would definitely like to see those numbers instead of just getting the final result table.

**Comment by loffleraSMH** [comment]
> > @guoyi-smh one more general comment/question (for a future sprint): Some of the filtering by date-time variables happens within the SQL query now, which probably makes it more efficient, but with this approach it’s not obvious how many date-time variables are missing/invalid. I’ve been adding a new utility function that transforms/checks date-time variables (see issue #76), so I think eventually, I’d like to apply this within these 3 functions as well so users are informed about missing/invalid date-times? This might be particularly relevant for imaging, so we can inform users how missing order date-times are handled. That would mean we’d query all entries first and then perform the filtering by date-times after. Any thoughts on this approach? If you agree, I can open a new issue for that so we can tackle that in a future update.
> 
> Agreed! I hesitated on this but chose the performance because I thought the numbers were not reported now anyways. As the user I would definitely like to see those numbers instead of just getting the final result table.

Ok sounds good, I'll go ahead and open a new issue for this so we can update it in the future.

File: R/n_routine_bloodwork.r [hunk 4]
```diff
@@ -55,58 +69,64 @@
...
-  ) %>%
-    as.data.table()
-
+  ) %>% as.data.table
 
+  # only count tests with a valid result_value (numeric or starting with "<", ">")
   lab <-
     lab[, result_value := .(trimws(result_value))] %>%
-    .[!is.na(as.numeric(result_value)) | startwith.any(result_value, c("<", ">"))] %>%
+    .[!is.na(as.numeric(result_value)) | grepl("^([<>])", result_value)] %>%
```
[the remark this answers is not in the document]
Comment (guoyi-smh) [reply R/n_routine_bloodwork.r:122]: Good point. And we have the same inconsistency issue in MPR. Currently in MPR the bloodwork indicator does not filter by whether the result is numeric or not, it counts all results. For the RBC transfusion indicator, it takes numeric values, but I just found that the code does not do what it claims to do:
```
# keep numeric values or those starting with <, >, @
  # exclude other non-numeric hgb result values
  hemoglobin[, result_value := as.numeric(
    stringr::str_split(result_value,pattern = "@|<|>",
                       simplify = TRUE)[,1])]
``` 
This code will keep any value before "<" or ">" or "@", which is right only for "@". For "<" or ">", we want to include the value following "<" or ">", not after. 

Here is what I do for my research projects and I just had a look at all Sodium and HGB test results it would capture most of the non-num result values except for some very rare edge cases. 
Generally the following 2 types should be included:
1. starting with ">", "<" or "LESS THAN", and followed by a numeric number
2. starting with a numeric number, followed by some string starting with "@", like `"138@WPDR"`, `"84@WPDR"`, `"129@"`. In this type we should only include the number before the "@" sign, not after. Results like `"TO CT @2137"`, "TO ER @1720" should be excluded.
In some rare cases it is the combination of the two, like `">180@"`.

I will fix the two functions here following this approach and create a new issue in MPR repo.

**Comment by guoyi-smh** [comment]
Hi @loffleraSMH ,  Thank you so much for you thorough  and thoughtful review! I have addressed the issues in the latest push. Please review the update when you have time. Thanks!!

File: R/n_rbc_transfusions.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (loffleraSMH) [reply R/n_rbc_transfusions.R:93]: No problem, thanks for fixing! I checked and confirmed that the 2 sites are now excluded. I'll go ahead and resolve this comment.

File: R/n_routine_bloodwork.r (hunk 4 above)
[the remark this answers is not in the document]
Comment (loffleraSMH) [reply R/n_routine_bloodwork.r]: Amazing, thanks so much @guoyi-smh for digging into this! That all makes sense to me, and thanks so much for ensuring consistency between MPR/OPR and Rgemini. I reviewed the updated code and everything looks good. I'll go ahead and resolve this conversation.

**Review by loffleraSMH — APPROVED** [review]
Thanks so much @guoyi-smh for implementing the additional changes! Everything looks good now. I approved the PR and you can go ahead and merge.
